docs: update `README` to reflect the `trafix-codec-v0.1.0` milestone (#25)
oxifix/trafix: README.md

# Project Structure

> This project is organized as a multirepo, consisting of three separate crates:

Reviewer: I would suggest slightly less specific language to make this correct when we change project structure as well.

```suggestion
This project is organized as a multirepo, consisting of multiple underlying crates, some of which are:
```

3. **trafix** — an umbrella crate wrapping the 2 below.
1. **trafix-codec** — a low-level library for high-performance parsing, encoding, and validation of FIX messages.  

> 2. **trafix-engine** — a full-featured FIX engine with session management, transports, and persistence built on top of `trafix-codec`.

Reviewer: ```suggestion
- **trafix** — an umbrella crate wrapping the 2 below.
- **trafix-codec** — a low-level library for high-performance parsing, encoding, and validation of FIX messages.  
- **trafix-engine** — a full-featured FIX engine with session management, transports, and persistence built on top of `trafix-codec`.
```

## trafix-codec

**trafix-codec** is the foundation crate: it provides control over FIX messages manipulation.

### Key Features

> - **Message creation**: Version-agnostic FIX message creation using a safe builder.

Reviewer: ```suggestion
- **Message creation**: Version-agnostic FIX message creation using a type-safe builder.
```

  - Generic builder for constructing Message instances.
  - Allowing of custom fields addition.

>   - Forbidding of message building until all the required fields are provided.

Reviewer: ```suggestion
  - Type-safe builder for `Message`.
  - Allows custom fields.
  - Prevents construction of invalid `Message`s.
```

> - **Message decoding**: Version-agnostic FIX message decoding with a focus on correctness and performance.

Reviewer: We do not focus on performance as much at the moment. 

```suggestion
- **Message decoding**: Version-agnostic FIX message decoding with a focus on correctness.
```

  - SOH-based frame parsing (`tag=value\x01`)
  - Support for FIX session-level messages

>   - Validation of `BeginString (8)`, `BodyLength (9)`, and `CheckSum (10)`

Reviewer: I would suggest to be less specific, we support more than just this. 

```suggestion
  - Support for message validation
```

- **Message encoding**: Deterministic FIX message serialization with automatic framing.
  - Correct handling of `BeginString (8)`, `BodyLength (9)`, and `CheckSum (10)`
  - Supports arbitrary field order while enforcing FIX wire rules

>   - Efficient buffer-based encoding (`Bytes` / `BytesMut`)

Reviewer: ```suggestion
  - Efficient buffer-based encoding using (`Bytes` / `BytesMut`)
```

---

### Getting Started

Add `trafix-codec` to your `Cargo.toml`:

```toml
[dependencies]
trafix-codec = "0.1.0"
```

---

> 

Reviewer: I suggest removing this part because it's obvious (we don't use any specific features etc.)

### Example: full usage

```rust
use trafix_codec::message::{
    Message,
    field::{
        Field,
        value::{begin_string::BeginString, msg_type::MsgType},
    },
};

fn main() {
    // Create the FIX message builder
    let builder = Message::builder(BeginString::FIX44, MsgType::Logout);

    // Create a custom field (to be added in header)
    let custom_header_field = Field::Custom {
        tag: 22,
        value: b"custom_header_field".to_vec(),
    };

    // Create a custom field (to be added in body)
    let custom_body_field1 = Field::Custom {
        tag: 40000,
        value: b"custom_body_field1".to_vec(),
    };

    // Create a custom field (to be added in body)
    let custom_body_field2 = Field::Custom {
        tag: 50000,
        value: b"custom_body_field2".to_vec(),
    };

    // build the message after adding the needed fields.
    let message = builder
        .with_header(custom_header_field.clone())
        .with_field(custom_body_field1.clone())

>         .with_field(custom_body_field2.clone())

Reviewer: Why do we need to clone here?

        .build();

    // encode the message into bytes
    let encoded_message = message.encode();
    println!("{:?}", encoded_message);

>     // decoding example; decoing and encoding a message results in the same bytes output.

Reviewer: ```suggestion
    // decoding example; decoding and encoding a message results in the same bytes output.
```
